core/src/main/scala/cats/data/StateT.scala

```diff
@@ -1,75 +1,90 @@
 package cats
 package data
 
+import cats.functor.{ Contravariant, Profunctor }
 import cats.syntax.either._
 
 /**
- * `StateT[F, S, A]` is similar to `Kleisli[F, S, A]` in that it takes an `S`
- * argument and produces an `A` value wrapped in `F`. However, it also produces
- * an `S` value representing the updated state (which is wrapped in the `F`
- * context along with the `A` value.
+ *
+ * `IndexedStateT[F, SA, SB, A]` is a stateful computation in a context `F` yielding
+ * a value of type `A`. Its state transitions from a value of type `SA` to a value
+ * of type `SB`.
+ *
+ * Given `IndexedStateT[F, S, S, A]`, this yields the plain `StateT[F, S, A]`.
+ *
+ * Note that `IndexedStateT[F, SA, SB, A]` is not a monad, but an indexed monad.
  */
-final class StateT[F[_], S, A](val runF: F[S => F[(S, A)]]) extends Serializable {
+final class IndexedStateT[F[_], SA, SB, A](val runF: F[SA => F[(SB, A)]]) extends Serializable {
 
-  def flatMap[B](fas: A => StateT[F, S, B])(implicit F: FlatMap[F]): StateT[F, S, B] =
-    StateT.applyF(F.map(runF) { sfsa =>
-      sfsa.andThen { fsa =>
-        F.flatMap(fsa) { case (s, a) =>
-          fas(a).run(s)
+  def flatMap[B, SC](fas: A => IndexedStateT[F, SB, SC, B])(implicit F: FlatMap[F]): IndexedStateT[F, SA, SC, B] =
+    IndexedStateT.applyF(F.map(runF) { safsba =>
+      safsba.andThen { fsba =>
+        F.flatMap(fsba) { case (sb, a) =>
+          fas(a).run(sb)
         }
       }
     })
 
-  def flatMapF[B](faf: A => F[B])(implicit F: FlatMap[F]): StateT[F, S, B] =
-    StateT.applyF(F.map(runF) { sfsa =>
+  def flatMapF[B](faf: A => F[B])(implicit F: FlatMap[F]): IndexedStateT[F, SA, SB, B] =
+    IndexedStateT.applyF(F.map(runF) { sfsa =>
       sfsa.andThen { fsa =>
         F.flatMap(fsa) { case (s, a) => F.map(faf(a))((s, _)) }
       }
     })
 
-  def map[B](f: A => B)(implicit F: Functor[F]): StateT[F, S, B] =
+  def map[B](f: A => B)(implicit F: Functor[F]): IndexedStateT[F, SA, SB, B] =
     transform { case (s, a) => (s, f(a)) }
 
+  def contramap[S0](f: S0 => SA)(implicit F: Monad[F]): IndexedStateT[F, S0, SB, A] =
```

Comment:
doesn't this only need `Applicative[F]`?

`(pure(f), runF).map2(_.andThen(_))` I *think* does it.